WIP- card draw animation

AllanChen-eng opened the pull request:
Pushing early with game scene changes to avoid merge conflicts
Still a work in progress, to do:
- improve the animation
-delay when card is added to your hand (should wait for drawing motion)
- add opposing player card draw
- interact with UI canvas

AllanChen-eng committed:
WIP- card draw animation

Made necessary game scene changes to avoid merge conflicts
Still a work in progress, to do:
- improve the animation
-delay when card is added to your hand (should wait for drawing motion)
- add opposing player card draw
- interact with UI canvas

arjohnston commented on Assets/Scripts/Cards/DeckManager.cs:
```diff
@@ -149,8 +149,10 @@ public class DeckManager : NetworkSingleton<DeckManager> {
     public void DrawInitialHandOfCards() {
-        for (int i = 0; i < GameSettings.initialPlayerHandSize; i++) {
-            DrawCard();
+        //Didn't want to use update since this is only called once
+        for (float i = GameSettings.initialPlayerHandSize; i > 0; i--) {
+            Invoke("DrawCard",(i-(.5f*i)));
+            Debug.Log("invoked at" + (i - (.5f*i)));
```
Can you instead pass in a boolean to make this a little cleaner? Something like: `public void DrawCard(bool shouldAnimate = true)`, then you can update the this declaration with `DrawCard(false)`

arjohnston commented on Assets/Scripts/Cards/CardDrawAnimation.cs:
```diff
@@ -0,0 +1,77 @@
+    private bool[] animating;
+    private GameObject[] deck;
+    private int animationsToDo=0;
+    private int totalAllowedAnimations = 10; // 10 cause 5x2 for start of game card draws
```
Hmm, I thought we aren't animating the start? This smells like a magic number

arjohnston commented on Assets/Scripts/Cards/CardDrawAnimation.cs:
```diff
@@ -0,0 +1,77 @@
+    // Start is called before the first frame update
+    private Vector3 p1DrawStartLocation;
+    private Vector3 p2DrawStartLocation;
+    public GameObject playerHandArea;
```
I think you can call `PlayerHand.Instance.GetPlayerHandArea` or something - if it doesn't exist, I would make one. This way you don't need to have multiple components with the same UI element attached (can get messy when altering elements)

arjohnston commented on Assets/Scripts/Cards/CardDrawAnimation.cs:
```diff
@@ -0,0 +1,77 @@
+    private Vector3 p1DrawStartLocation;
+    private Vector3 p2DrawStartLocation;
+    public GameObject playerHandArea;
+    public GameObject gameboardDeckPrefab;
```
Same here as above with playhand owning this

arjohnston commented on Assets/Scenes/Game.unity:
```diff
@@ -8052,6 +8053,20 @@ MonoBehaviour:
   playerTwoTrapsPlayed: {fileID: 1340026077}
   playerTwoSpellsPlayed: {fileID: 309575274}
   playerTwoEnchantmentsPlayed: {fileID: 122173697}
+--- !u!114 &983941261
+MonoBehaviour:
+  m_ObjectHideFlags: 0
+  m_CorrespondingSourceObject: {fileID: 0}
+  m_PrefabInstance: {fileID: 0}
+  m_PrefabAsset: {fileID: 0}
+  m_GameObject: {fileID: 983941246}
+  m_Enabled: 1
+  m_EditorHideFlags: 0
+  m_Script: {fileID: 11500000, guid: 6c3fde751b7354aec961a2126060cd79, type: 3}
+  m_Name: 
+  m_EditorClassIdentifier: 
+  playerHandArea: {fileID: 51128012}
+  gameboardDeckPrefab: {fileID: 1466029038368381171, guid: fdfd357dc1a5f41479ecbc204263d6aa, type: 3}
```
What was edited here?

arjohnston commented on Assets/Scripts/Cards/CardDrawAnimation.cs:
```diff
@@ -0,0 +1,77 @@
+    private int totalAllowedAnimations = 10; // 10 cause 5x2 for start of game card draws
+    void Start()
+    {
+        animating = new bool[totalAllowedAnimations];
+        animationProgress = new float[totalAllowedAnimations];
+        deck = new GameObject[totalAllowedAnimations];
+       p1DrawStartLocation = Gameboard.Instance.GetPlayerOneDeckArea().transform.position;
+       p2DrawStartLocation = Gameboard.Instance.GetPlayerTwoDeckArea().transform.position;
```
I would put these in an `Awake` method instead, so there are no race conditions

arjohnston commented on Assets/Scripts/Cards/CardDrawAnimation.cs:
```diff
@@ -0,0 +1,77 @@
+                        Debug.Log("locations match!");
+                    }
+                    animating[a] = false;
+                    Destroy(deck[a]);
```
I saw that you had indicated you want to have p2 draw cards and have that animate, too. I would probably recommend against that as that means we need a lot of Server/Client RPC calls for little gain. Also kind of difficult to tell where the P2 card should animate to. 

If you did want P2 animation, then this would need to be called on the server

AllanChen-eng committed:
Card draw animation clean up and UI when deck is clicked on

Clicking on the deck hex now shows the deck class and the remaining cards!
Cleaned up code from wip
- Game scene changes:
created textmeshUI on canvas
reference to script
refactored so references are called from class instances that use prefab
